ddtrace/sampler.py

```diff
@@ -44,59 +46,67 @@ def set_sample_rate(self, sample_rate):
         self.sampling_id_threshold = sample_rate * MAX_TRACE_ID
 
     def sample(self, span):
-        span.sampled = ((span.trace_id * KNUTH_FACTOR) % MAX_TRACE_ID) <= self.sampling_id_threshold
-        span.set_metric(SAMPLE_RATE_METRIC_KEY, self.sample_rate)
+        sampled = ((span.trace_id * KNUTH_FACTOR) % MAX_TRACE_ID) <= self.sampling_id_threshold
 
-class ThroughputSampler(object):
-    """ Sampler applying a strict limit over the trace volume.
+        return sampled
 
-        Stop tracing once reached more than `tps` traces per second.
-        Computation is based on a circular buffer over the last
-        `BUFFER_DURATION` with a `BUFFER_SIZE` size.
+class RateByServiceSampler(object):
+    """Sampler based on a rate, by service
 
-        DEPRECATED: Outdated implementation.
+    Keep (100 * `sample_rate`)% of the traces.
+    The sample rate is kept independently for each service/env tuple.
     """
 
-    # Reasonable values
-    BUCKETS_PER_S = 10
-    BUFFER_DURATION = 2
-    BUFFER_SIZE = BUCKETS_PER_S * BUFFER_DURATION
+    def __init__(self, sample_rate=1):
+        self._lock = Lock()
+        self._by_service_samplers = {}
+        self._default_key = self._key(None, None)
```

Comment:
OK, will refactor with `()` indeed :+1:. The use cases for using the default keys can be:

- a broken (custom or in-house) instrumentation which is now setting the service early enough. In that case it has no way to guess what rate to apply, because when they are submitted later, the spans do have a service set. So the idea here is to pick the default value returned by the agent, which is the global sampling rate. Technically, it causes over-sampling (dropping too many traces), which is not great, but not as bad as using `1` and send everything.
- an application that is starting and/or is having some service name change (services could be defined dynamically according to some variable flag, who knows, or some services could show up rarely and appear "some time" after the other) so when a new service shows up, instead of using 1 we use the default sampling rate (this is done here: https://github.com/DataDog/dd-trace-py/pull/359/files#diff-073d17b74132da4920e45dbb8d5de860R88) which at least is a little context dependant, if we're already at `0.01` we won't "jump" to `1` just because the service is new.

Globally, it's really about providing a default behavior that is a little context dependant, it's not the right rate (this, we only now later) but at least we get a reasonable hint on what to do.

Reply:
https://github.com/DataDog/dd-trace-py/pull/359/commits/10eac176d05725d5450c2c5bfa8a1acd7f002224